src/components/TooltipNew/UserGuideTooltip.js

```diff
@@ -0,0 +1,173 @@
+import memoizeOne from 'memoize-one';
+import { throttle } from '@livechat/data-utils';
+import styles from './style.scss';
+import PopperTooltip from './PopperTooltip';
+import ModalPortal from '../Modal/ModalPortal';
+import SpotlightOverlay from './SpotlightOverlay';
+import VirtualReference from '../../helpers/virtual-element-reference';
+
+const baseClass = 'guide-tooltip';
+
+const cx = classNames.bind(styles);
+
+const memoizedReference = memoizeOne(
+  (element, padding) => new VirtualReference(element, padding)
+);
+
+class UserGuideTooltip extends React.PureComponent {
+  static getDerivedStateFromProps(props, state) {
+    if (props.isVisible) {
+      if (!state.lastElement) {
+        // this is the first time the tooltip is shown - don't slide from the (0, 0) origin
+        return {
+          shouldSlide: false,
+          lastElement: props.element
+        };
+      }
+
+      if (props.element !== state.lastElement) {
+        // the element has changed - slide to the next one
+        return {
+          shouldSlide: true,
+          lastElement: props.element
+        };
+      }
+
+      // window was scrolled or any other change has happened - don't slide
+      return {
+        shouldSlide: false
+      };
+    }
+
+    // reset the element when the tooltip hides
+    return {
+      lastElement: null
+    };
+  }
+
+  constructor(props) {
+    super(props);
+
+    // we don't use the following state params but we still need to re-render when the window changes
+    /* eslint-disable react/no-unused-state */
+    this.state = {
+      windowWidth: 0,
+      windowHeight: 0,
+      windowScrollX: 0,
+      windowScrollY: 0
+    };
+    /* eslint-enable react/no-unused-state */
+
+    this.handleDocumentChange = throttle(
+      16,
+      this.handleDocumentChange.bind(this)
+    );
+  }
+
+  componentDidMount() {
+    window.addEventListener('resize', this.handleDocumentChange);
+    document.addEventListener('scroll', this.handleDocumentChange);
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('resize', this.handleDocumentChange);
+    document.removeEventListener('scroll', this.handleDocumentChange);
+  }
+
+  handleDocumentChange() {
+    const windowWidth = window.innerWidth;
+    const windowHeight = window.innerHeight;
+    const windowScrollX = document.documentElement.scrollLeft;
+    const windowScrollY = document.documentElement.scrollTop;
+
+    /* eslint-disable react/no-unused-state */
+    this.setState({
+      windowWidth,
+      windowHeight,
+      windowScrollX,
+      windowScrollY
+    });
+    /* eslint-enable react/no-unused-state */
+  }
+
+  render() {
+    const {
+      children,
+      className,
+      placement,
+      zIndex,
+      element,
+      isVisible,
+      slide,
+      theme,
+      containerName
+    } = this.props;
+    const referenceElement = memoizedReference(element, 8);
+    const rect = referenceElement.getBoundingClientRect();
```

Comment:
`getBoundingClientRect` on each render is not a good idea imho

Reply:
I added more memoization